src/cluster.c

```diff
@@ -2941,6 +2944,64 @@ int clusterProcessPacket(clusterLink *link) {
             decrRefCount(channel);
             decrRefCount(message);
         }
+    } else if (type == CLUSTERMSG_TYPE_MPUBLISH) {
+        if (!sender) return 1;  /* We don't know that node. */
+
+        const uint8_t *src, *end;
+        robj *channel, **messages;
+        uint32_t channel_len, message_len, len;
+        unsigned i, msg_count;
+
+        /* Don't bother creating useless objects if there are no
+         * Pub/Sub subscribers. */
+        if ((type == CLUSTERMSG_TYPE_MPUBLISH
+            && serverPubsubSubscriptionCount() > 0))
+        {
+            channel_len = ntohl(hdr->data.publish.msg.channel_len);
+            message_len = ntohl(hdr->data.publish.msg.message_len);
+            /* Count messages */
+            src = hdr->data.publish.msg.bulk_data + channel_len;
+            end = src + message_len;
+            msg_count = 0;
+            while (src + 4 <= end) {
+                memcpy(&len, src, sizeof(len));
+                len = ntohl(len);
+                src += 4;
+                if (src + len > end) {
+                    serverLog(LL_WARNING,
+                        "Received %s packet with malformed messages",
+                        clusterGetMessageTypeString(type));
+                    return 1;
+                }
+                src += len;
+                ++msg_count;
+            }
+            serverAssert(src <= end);
+            if (src != end) {
+                serverLog(LL_WARNING,
+                    "Received %s packet with malformed messages (short)",
+                    clusterGetMessageTypeString(type));
+                return 1;
+            }
+            channel = createStringObject(
+                        (char*)hdr->data.publish.msg.bulk_data,channel_len);
+            /* Parse them out */
+            messages = zmalloc(sizeof(messages[0]) * msg_count);
+            src = hdr->data.publish.msg.bulk_data + channel_len;
+            for (i = 0; src < end; ++i) {
+                memcpy(&len, src, sizeof(len));
+                len = ntohl(len);
+                src += 4;
+                messages[i] = createStringObject((char *) src, len);
+                src += len;
+            }
+            serverAssert(msg_count == i);
+            pubsubPublishMessages(channel, messages, (int) msg_count, 0);
```

Comment:
Maybe we can just loop over the messages and publish them individually. We would lesser memory consumption with this.